### PR TITLE
patch nested traces of the same type

### PR DESCRIPTION
**Pull request by logan-markewich**
# Description

When the callback manager traces nested entry points of the same trace id (i.e. query that has nested queries), the callback manager is unstable and may crash.

Now, nested trace ids are possible.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Future Plans

Async is still wonky, and this is a known issue, should be fixed. One possible path to a solution is creating a dispatcher to properly aggregate trace maps from async functions.

### CODE REVIEW

File: llama_index/callbacks/base.py [hunk 1]
```diff
@@ -135,29 +135,34 @@ def as_trace(self, trace_id: str) -> Generator[None, None, None]:
...
-        if trace_id is not None and trace_id == self._trace_id:
-            for handler in self.handlers:
-                handler.end_trace(trace_id=trace_id, trace_map=self._trace_map)
-            self._trace_id = None
+        if trace_id is not None and self._trace_id_stack is not None:
+            self._trace_id_stack.pop()
+            if len(self._trace_id_stack) == 0:
+                for handler in self.handlers:
+                    handler.end_trace(trace_id=trace_id, trace_map=self._trace_map)
+                self._trace_id_stack = None
```
Comment (jerryjliu) [inline llama_index/callbacks/base.py:160]: should we throw an error if trace id stack is None / should that ever happen here?

File: llama_index/callbacks/base.py [hunk 2]
```diff
@@ -135,29 +135,34 @@ def as_trace(self, trace_id: str) -> Generator[None, None, None]:
...
         self,
         trace_id: Optional[str] = None,
         trace_map: Optional[Dict[str, List[str]]] = None,
     ) -> None:
         """Run when an overall trace is exited."""
-        if trace_id is not None and trace_id == self._trace_id:
-            for handler in self.handlers:
-                handler.end_trace(trace_id=trace_id, trace_map=self._trace_map)
-            self._trace_id = None
+        if trace_id is not None and self._trace_id_stack is not None:
```
Comment (jerryjliu) [inline llama_index/callbacks/base.py:155]: should we throw an error if trace id stack is None / should that ever happen here?

File: llama_index/callbacks/base.py [hunk 3]
```diff
@@ -76,8 +76,8 @@ def __init__(self, handlers: List[BaseCallbackHandler]):
         """Initialize the manager with a list of handlers."""
         self.handlers = handlers
         self._trace_map: Dict[str, List[str]] = defaultdict(list)
-        self._trace_stack: List[str] = [BASE_TRACE_ID]
-        self._trace_id: Optional[str] = None
+        self._trace_event_stack: List[str] = [BASE_TRACE_EVENT]
+        self._trace_id_stack: Optional[List[str]] = None
```
Comment (jerryjliu) [inline llama_index/callbacks/base.py:80]: is there a conceptual reason why _trace_id_stack would be optional type on `__init__` but `_trace_event_stack` is a list with one item on `__init__`? find it a tad confusing

**Review by jerryjliu — APPROVED** [review]
thanks! some comments but approving to unblock

File: llama_index/callbacks/base.py (hunk 2 above)
Comment (logan-markewich) [reply llama_index/callbacks/base.py:155]: It shouldn't ever happen here, but it's just a type checking symptom of the callback manager sharing a base class with the handlers

File: llama_index/callbacks/base.py (hunk 3 above)
Comment (logan-markewich) [reply llama_index/callbacks/base.py:80]: Early morning code I guess lol let me double check if it's fine to make it non optional type

**Pull request merged**